server/bin/basket-proxy-server.js

```diff
@@ -59,51 +73,84 @@ function basketRequest(path, method, params) {
       'X-API-Key': API_KEY
     },
     form: params
+  };
+
+  var deferred = p.defer();
+
+  request(config, function (err, httpRequest, body) {
+    if (err) {
+      logger.error('basket.error', err);
+      return deferred.reject(err);
+    }
+
+    if (httpRequest.statusCode !== 200) {
+      // TODO do something here.
+    }
+
+    deferred.resolve(body);
   });
 
-  return req;
+  return deferred.promise;
+}
+
+function forwardBasketResponse(res, path, method, params) {
+  return basketRequest(path, method, params)
+    .then(function (body) {
+      res.type('application/json');
+      res.send(body);
+    }, function (err) {
+      res.status(400).send(err);
+    });
 }
 
 function initApp() {
   var app = express();
   app.use(bodyParser.json());
 
-  app.get('/lookup-user', function (req, res) {
+  app.use(cors({
+    origin: CORS_ORIGIN
+  }));
 
-    verifyOAuthToken(req.headers.authorization)
-      .then(function (creds) {
-        basketRequest('/lookup-user?email=' + creds.email, 'get').pipe(res);
-      });
-  });
+  app.get('/lookup-user', verifyOAuthToken(function (creds, req, res) {
+    forwardBasketResponse(res, '/lookup-user/?email=' + creds.email, 'get');
+  }));
 
-  app.post('/subscribe', function (req, res) {
+  app.post('/subscribe', verifyOAuthToken(function (creds, req, res) {
+    // Basket is slow, don't wait around for its response.
+    res.json({ status: 'ok' });
 
-    verifyOAuthToken(req.headers.authorization)
-      .then(function (creds) {
-        basketRequest('/subscribe', 'post', {
-          email: creds.email,
-          newsletter_id: req.params.newsletter_id,
-          lang: req.params.lang
-        }).pipe(res);
+    basketRequest('/subscribe/', 'post', {
+      email: creds.email,
+      newsletters: req.body.newsletters.join(','),
+      optin: 'Y'
     });
-  });
-
-  app.post('/unsubscribe', function (req, res) {
-
-    verifyOAuthToken(req.headers.authorization)
-      .then(function (creds) {
-        basketRequest('/unsubscribe', 'post', {
+  }));
+
+  app.post('/unsubscribe', verifyOAuthToken(function (creds, req, res) {
+    // Basket is slow, don't wait around for its response.
+    res.json({ status: 'ok' });
+
+    basketRequest('/lookup-user/?email=' + creds.email, 'get')
+      .then(function (body) {
+        var token;
+        try {
+          token = JSON.parse(body).token;
+        } catch(e) {
+          return res.status(400).send('error');
+        }
+
+        basketRequest('/unsubscribe/' + token + '/', 'post', {
           email: creds.email,
-          newsletter_id: req.params.newsletter_id
-        }).pipe(res);
-    });
-  });
+          newsletters: req.body.newsletters.join(',')
```

Comment:
Do we expect the basket server to eventually perform this coercion or should the client format the `newsletters` field?